flink-runtime/src/main/java/org/apache/flink/runtime/jobClient/JobClientUtils.java

```diff
@@ -0,0 +1,257 @@
+			NetUtils.ipAddressToUrlString(InetAddress.getByName(address.host().get())) :
+			"(unknown)";
+		int port = address.port().isDefined() ? ((Integer) address.port().get()) : -1;
+		LOG.info("Started JobClientUtils actor system at " + hostAddress + ':' + port);
+
+		Timeout timeout = new Timeout(AkkaUtils.getClientTimeout(config));
+		return new AkkaRpcService(system, timeout);
+	}
+
+	/**
+	 * Attaches to a running Job using the JobID, and wait for its job result
+	 *
+	 * @param jobID                  id of job
+	 * @param jobMasterGateway       gateway to the JobMaster
+	 * @param rpcService
+	 * @param leaderRetrievalService leader retriever service of jobMaster
+	 * @param sysoutLogUpdates       whether status messages shall be printed to sysout
+	 * @param timeout                register timeout
+	 * @param configuration          the flink configuration
+	 * @return
+	 * @throws JobExecutionException
+	 */
+	public static JobExecutionResult retrieveRunningJobResult(
+		JobID jobID,
+		JobMasterGateway jobMasterGateway,
+		RpcService rpcService,
+		LeaderRetrievalService leaderRetrievalService,
+		boolean sysoutLogUpdates,
+		FiniteDuration timeout,
+		Configuration configuration) throws JobExecutionException
+	{
+
+		checkNotNull(jobID, "The jobID must not be null.");
+		checkNotNull(jobMasterGateway, "The jobMasterGateway must not be null.");
+		checkNotNull(rpcService, "The rpcService must not be null.");
+		checkNotNull(leaderRetrievalService, "The leaderRetrievalService must not be null.");
+		checkNotNull(timeout, "The timeout must not be null");
+		checkNotNull(configuration, "The configuration must not be null");
+
+		JobInfoTracker jobInfoTracker = null;
+		try {
+			jobInfoTracker = new JobInfoTracker(rpcService, leaderRetrievalService, jobID, sysoutLogUpdates);
+			jobInfoTracker.start();
+			registerClientAtJobMaster(jobID, jobInfoTracker.getAddress(), jobMasterGateway, timeout);
+			ClassLoader classLoader = retrieveClassLoader(jobID, jobMasterGateway, configuration);
+			return awaitJobResult(jobInfoTracker, classLoader);
+		} finally {
+			if (jobInfoTracker != null) {
+				jobInfoTracker.shutDown();
+			}
+		}
+	}
+
+	/**
+	 * Awaits the result of the job execution which jobInfoTracker listen for
+	 *
+	 * @param jobInfoTracker job info tracker
+	 * @param classLoader    classloader to parse the job result
+	 * @return
+	 * @throws JobExecutionException
+	 */
+	public static JobExecutionResult awaitJobResult(JobInfoTracker jobInfoTracker,
+		ClassLoader classLoader) throws JobExecutionException
+	{
+		try {
+			while (true) {
+				Future<JobExecutionResult> jobExecutionResultFuture = jobInfoTracker.getJobExecutionResult(classLoader);
+				try {
+					JobExecutionResult jobExecutionResult = Await.result(jobExecutionResultFuture, new Timeout(AkkaUtils.INF_TIMEOUT()).duration());
+					return jobExecutionResult;
+				} catch (TimeoutException e) {
+					// ignore timeout exception, retry
```

Comment:
I'm not sure if that will cause an infinite loop in case of reoccurring connection timeouts.